docs/rfcs/021-raii-runtime.md

```diff
@@ -0,0 +1,209 @@
+# Improving runtime by making use of the RAII idiom
+
+## Motivation
+
+Currently, we have several problems regarding the runtime:
+
+- `TenantState` doesn't clearly show what happens with the tenant. Repository and code comments need to describe how state transitions occur and what they mean. It is critical since the code is going to have more states. The same applies to `TimelineState`.
+- Current tenant and timeline access, shutdown design fails to shut down tenants properly, timelines, background tasks, and control their lifetimes, resulting in races.
+- `tasks_mgr` runtime is redundant, and its design allows us to have even more races without noticing.
+
+This RFC describes how to rewrite a runtime in a more RAII fashion, fixing part of its problems.
+
+## Small quality of life changes
+
+Before we start, some synchronisation primitives should be proposed.
+
+1. In our code, there're multiple places where we use `tokio::watch::{Sender, Receiver}` with type `()`. It's hard to track down whether we're just waiting for the drop, for a signal or both from the sender. To make it clear, a couple of wrappers are needed:
+    - `SignalSender` / `SignalWatcher`: wait only for proper signals and `panic!` when it receives `RecvError`.
```

Comment:
IMO, this should mean we broke a core invariant (`SignalSender ` couldn't be dropped before receivers drop) when coding. This wrapper's point is to show the engineer this couldn't happen when he sees the type. We should stop at least the tenant and its tasks immediately and, even better - panic.

Reply:
Can we somehow convert this to compile time error? Either via some type-state pattern or by using Undroppable type from @problame's experiments. In short it generates drop impl that points to non existing symbol so linker will emit an error if this code ends up included in the binary

Reply:
We can't since the invariant _"is `Sender` dropped after `Receiver`s?_" could only be checked in runtime. This means we can only change this to just non-generic regular `Sender` / `Receiver`, but it still won't show the user that we expect the invariant to hold in runtime. It should be considered a serious disaster, a bug potentially leading to races and other severe consequences, because engineers can rely on it.

Anyway, I removed that from RFC because I couldn't find a great example from `SignalSender` / `SignalReceiver` in our code. There's [timeline layer flush](https://github.com/neondatabase/neon/blob/main/pageserver/src/tenant/timeline.rs#L140), but it needs another synchronisation primitive, and [shutdown_rx](https://github.com/neondatabase/neon/blob/main/pageserver/src/task_mgr.rs#L153), but I think we'll benefit from removing it. Perhaps, the idea of having such a wrapper is not good.